sbi/utils/conditional_density.py

```diff
@@ -321,3 +325,376 @@ def _normalize_probs(probs: Tensor, limits: Tensor) -> Tensor:
+        dist = self.net._distribution
+        logits, m, prec, *_ = dist.get_mixture_components(encoded_x)
+        norm_logits = logits - torch.logsumexp(logits, dim=-1, keepdim=True)
+
+        scale = self.net._transform._transforms[0]._scale
+        shift = self.net._transform._transforms[0]._shift
+
+        self.mc = torch.exp(norm_logits).detach()
+        self.m = ((m - shift) / scale).detach()[0]
+
+        L = torch.cholesky(
+            prec[0].detach() + torch.eye(self.m.shape[1]) * 1e-6
+        )  # sometimes matrices are not pos semi def. dirty fix.
+        C = torch.inverse(L)
+        self.S = C.transpose(2, 1) @ C
+        A_inv = torch.inverse(scale * torch.eye(self.S.shape[1]))
+        self.S = A_inv @ self.S @ A_inv.T
+
+        return self.mc, self.m, self.S
+
+    def log_prob(self, X: Tensor, individual=False) -> Tensor:
+        """Evaluates the Mixture of Gaussian (MoG)
+        probability density function at a value x.
+
+        Adpated from code courtesy of @ybernaerts.
+
+        Args:
+            X: Values at which to evaluate the MoG pdf.
+            individual: If True the probability density is returned for each cluster component.
+
+        Returns:
+            log_prob: Log probabilities at values specified by X.
+        """
+
+        pdf = torch.zeros((X.shape[0], self.m.shape[0]))
+        for i in range(self.m.shape[0]):
+            pdf[:, i] = self.mulnormpdf(X, self.m[i], self.S[i]) * self.mc[0, i]
+        if individual:
+            return torch.log(pdf)
+        else:
+            log_factor = torch.log(self.leakage_correction(x=self.default_x))
+            return torch.log(torch.sum(pdf, axis=1)) - log_factor
+
+    def sample(self, sample_shape: Tuple[int, int]) -> Tensor:
+        """Draw samples from a Mixture of Gaussians (MoG)
+
+        Adpated from code courtesy of @ybernaerts.
+        Args:
+            sample_shape: The number of samples to draw from the MoG distribution.
+
+        Returns:
+            X: A matrix with samples rows, and input dimension columns.
+        """
+
+        K, D = self.m.shape  # Determine dimensionality
+
+        num_samples = torch.Size(sample_shape).numel()
+        pbar = tqdm(
+            total=num_samples,
+            desc=f"Drawing {num_samples} posterior samples",
+        )
+
+        # Cluster selection
+        cs_mc = torch.cumsum(self.mc, 1)
+        cs_mc = torch.hstack((torch.tensor([[0]]), cs_mc))
+        sel_idx = torch.rand(num_samples)
+
+        # Draw samples
+        res = torch.zeros((num_samples, D))
+        f1 = sel_idx[:, None] >= cs_mc
+        f2 = sel_idx[:, None] < cs_mc
+        idxs = f1[:, :-1] * f2[:, 1:]
+        ksamples = torch.sum(idxs, axis=0)
+        for k, samplesize in enumerate(ksamples):
+            # draw initial samples
+            multivar_normal = torch.distributions.multivariate_normal.MultivariateNormal(
+                self.m[k], self.S[k]
+            )
```

Comment:
We should avoid creating objects in a loop and instead sample by multiplying standard normal samples with the cholesky factor.